cpp/src/io/json/write_json.cu

```diff
@@ -0,0 +1,743 @@
+#include <cudf/structs/struct_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <thrust/distance.h>
+#include <thrust/execution_policy.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/scan.h>
+#include <thrust/tabulate.h>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace cudf::io::json::detail {
+
+using namespace cudf::io;
+std::unique_ptr<column> make_column_names_column(host_span<column_name_info const> column_names,
+                                                 size_type num_columns,
+                                                 rmm::cuda_stream_view stream);
+namespace {
+
+/**
+ * @brief Functor to modify a string column for JSON format.
+ *
+ * This will convert escape characters and wrap quotes around strings.
+ */
+struct escape_strings_fn {
+  column_device_view const d_column;
+  string_view const d_delimiter;  // check for column delimiter
+  offset_type* d_offsets{};
+  char* d_chars{};
+
+  __device__ void write_char(char_utf8 chr, char*& d_buffer, offset_type& bytes)
+  {
+    if (d_buffer)
+      d_buffer += cudf::strings::detail::from_char_utf8(chr, d_buffer);
+    else
+      bytes += cudf::strings::detail::bytes_in_char_utf8(chr);
+  }
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_column.is_null(idx)) {
+      if (!d_chars) d_offsets[idx] = 0;
+      return;
+    }
+
+    auto const d_str = d_column.element<string_view>(idx);
+
+    // entire string must be double-quoted.
+    constexpr char_utf8 const quote = '\"';  // wrap quotes
+    bool constexpr quote_row        = true;
+
+    char* d_buffer    = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    offset_type bytes = 0;
+
+    if (quote_row) write_char(quote, d_buffer, bytes);
+    for (auto chr : d_str) {
+      auto escaped_chars = cudf::io::json::experimental::detail::get_escaped_char(chr);
+      if (escaped_chars.first == '\0') {
+        write_char(escaped_chars.second, d_buffer, bytes);
+      } else {
+        write_char(escaped_chars.first, d_buffer, bytes);
+        write_char(escaped_chars.second, d_buffer, bytes);
+      }
+    }
+    if (quote_row) write_char(quote, d_buffer, bytes);
+
+    if (!d_chars) d_offsets[idx] = bytes;
+  }
+};
+
+// Struct concatenation.
+struct concat_structs_base {
+  table_device_view const d_table;
+  column_device_view const d_column_names;
+  string_view const prepend_row;            //{
+  string_view const append_row;             //} or }\n for json-lines
+  string_view const d_col_separator;        //:
+  string_view const d_val_separator;        //,
+  string_scalar_device_view const d_narep;  // null
+  bool include_nulls;
```

Comment:
Typically, I avoid assigning defaults to leaf level functions. Defaults come from top level arguments. Here it comes from json_writer_options.
I updated anyway to `false`.

Reply:
Including defaults in the declaration prevents UB in case there's a gap in the initialization (e.g. json_writer_options) and the data member is unintentionally left uninitialized. Easier to find the bug if `include_nulls` doesn't flip-flop randomly.